media-sound/zynaddsubfx: build new UI instead of unmaintained one

zx2c4 opened the pull request:
ZynAddSubFX hasn't used FLTK for years. It sits in the repo but is basically unmaintained. Instead there's a new UI based on something called "Zest". The build system is horrific. The upstream author charges $45 for a prebuilt binary but keeps the sources as GPLv2.

Thanks to [the amazing work](https://gitlab.archlinux.org/archlinux/packaging/packages/zynaddsubfx/-/blob/main/PKGBUILD?ref_type=heads) of @dvzrv we have an idea of how to build this ourselves. This patch is based directly on David's PKGBUILD.

This ebuild was already in pretty bad shape, and this patch here takes the path of least resistance toward getting it updated to use the new UI. But there's a lot wrong with it still, and this patch does some gnarly things. It might require a lot of love to get up to snuff, but at least this patch brings things into 2023.

Feel free to take this patch and modify it as you see fit to get things building. I'd prefer that the ebuild maintainer just take this patch over, rather than a tedious patch review process.

I didn't rename the -r1 to -r2 so that the github diff is easier to read, but obviously that should be done before merging.

zx2c4 commented:
CC @vimproved @fordfrog @a17r

zx2c4 commented:
https://bugs.gentoo.org/920178

vimproved commented:
Hey! Thanks so much for doing this, I was intending to clean this up and submit it at some point but never got around to it. Unfortunately, I think this is going to be gnarly no matter what we do because of the typical linux audio app stuff, but the important thing is that it works!

vimproved commented on media-sound/zynaddsubfx/zynaddsubfx-3.0.6-r1.ebuild:
```diff
@@ -42,11 +49,22 @@ BDEPEND="
 src_prepare() {
+	local i
+	for i in "${MRUBY_ZEST_PATCHES[@]}"; do
+		patch -d mruby-zest-build -Np1 -i "$i" || die
```
Instead of calling patch directly here, I'd do a pushd/popd into the directory and use eapply.

vimproved commented on media-sound/zynaddsubfx/zynaddsubfx-3.0.6-r1.ebuild:
```diff
@@ -7,12 +7,21 @@ inherit cmake flag-o-matic
+#   $ scp ../mruby-zest-build-${PV}.tar.gz dev.gentoo.org:public_html/distfiles/
+SRC_URI="
+	mirror://sourceforge/zynaddsubfx/${P}.tar.bz2
+	https://dev.gentoo.org/~zx2c4/distfiles/mruby-zest-build-${PV}.tar.gz
```
Repackaging this isn't necessary. The full UI sources including submodules are in `mirror://sourceforge/zynaddsubfx/zyn-fusion-ui-src-${PV}.tar.bz2`

zx2c4 replied:
Thanks!

vimproved commented on media-sound/zynaddsubfx/zynaddsubfx-3.0.6-r1.ebuild:
```diff
@@ -71,21 +89,33 @@ src_configure() {
+
+	cd mruby-zest-build
+	newbin zest zyn-fusion
+	insinto /usr/lib64/${PN}
```
`insinto /usr/$(get_libdir)/${PN}`

vimproved reviewed (changes requested):
Looks good overall except for the couple things I pointed out.

fordfrog commented on media-sound/zynaddsubfx/zynaddsubfx-3.0.6-r1.ebuild:
```diff
@@ -42,11 +49,22 @@ BDEPEND="
+	"${FILESDIR}"/${PN}-mruby-zest-build-${PV}-cflags_ldflags.patch
+	"${FILESDIR}"/${PN}-mruby-zest-build-${PV}-libzest_location.patch
+	"${FILESDIR}"/${PN}-mruby-zest-build-${PV}-makefile_find.patch
+	"${FILESDIR}"/${PN}-mruby-zest-build-${PV}-system_wide_location.patch
```
can't we just update those so that they can be applied using the standard PATCHES mechanism?

zx2c4 replied:
Not particularly easily, no, because with the src provided, the tarball's paths aren't very nice.

zx2c4 commented:
@fordfrog @vimproved I incorporated @vimproved's suggestions and cleaned up the patching a bit. If this looks okayish (and by "okayish" I mean "still bad because the ebuild is crusty but not worse than before), I'll rename the -r1 to -r2 and push it gentoo.git.

fordfrog commented:
i just tested the runtime and it seems to work fine so unless @vimproved has any more suggestions, we can imo merge it.

vimproved commented:
LGTM. If I wanted to nitpick I'd say scrub/add headers to the patches but that could be done on push.

fordfrog commented:
i scrubbed the patches, renamed the ebuild to -r2 and pushed this pr. thank you for this great improvement!

zx2c4 commented:
Thanks! Beat me to it.
